incorrect value returned for num_columns, causing OOB rows.get_value

```console
Runtime error: UNIQUE constraint failed: sqlite_autoindex_databases_1.hostname (19)
[1.139ms] [rows:0] INSERT INTO `databases` (`created_at`,`updated_at`,`deleted_at`,`hostname`,`namespace`,`fly_app`,`address`,`primary_address`,`cloud_cluster_name`,`local`,`allowed_ips`) VALUES ("2025-09-17 19:30:56.763","2025-09-17 19:30
:56.763",NULL,"test.local","ns-test","test-app","http://test:8080","","cluster-1",false,"") RETURNING `id`
    gorm_test.go:75: create failed: Runtime error: UNIQUE constraint failed: sqlite_autoindex_databases_1.hostname (19)
--- FAIL: TestGormBasicOperations (0.01s)


thread '<unnamed>' panicked at ~/Projects/turso/core/vdbe/mod.rs:1046:18:
internal error: entered unreachable code: a row should be formed of values only
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5, aborting
SIGABRT: abort
```

This is likely from the `database/sql` driver reporting the wrong amount of result rows, and `rows.get_row` being called when there is no value to be had.

--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,5 @@
 /target
 /_libs_artifacts
+*.db
+*.db-wal
+output.log

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,4 +14,4 @@ env_logger = { version = "0.11.6", default-features = false }
 tracing = "0.1.41"
 tracing-appender = "0.2.3"
 tracing-subscriber = {version =  "0.3.20", features = ["env-filter"] }
-turso_core = "0.1.5"
+turso_core = { version = "0.2.0-pre.6" }

--- a/go.mod
+++ b/go.mod
@@ -5,9 +5,18 @@ go 1.23.4
 require (
 	github.com/ebitengine/purego v0.8.3-0.20250507171810-1638563e3615
 	golang.org/x/sys v0.29.0
+	gorm.io/driver/sqlite v1.6.0
+	gorm.io/gorm v1.31.0
+)
+
+require (
+	github.com/jinzhu/inflection v1.0.0 // indirect
+	github.com/jinzhu/now v1.1.5 // indirect
+	github.com/mattn/go-sqlite3 v1.14.22 // indirect
+	golang.org/x/text v0.20.0 // indirect
 )
 
 retract (
-	v0.0.1
 	v0.1.4
+	v0.0.1
 )

--- a/gorm_test.go
+++ b/gorm_test.go
@@ -203,6 +203,42 @@ func TestGormUpsertWithReturning(t *testing.T) {
 	})
 }
 
+func TestGormSoftDelete(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+	testData := []Database{
+		{Hostname: "soft1.local", Namespace: "ns1"},
+		{Hostname: "soft2.local", Namespace: "ns2"},
+		{Hostname: "soft3.local", Namespace: "ns3"},
+	}
+	err = db.Create(&testData).Error
+	if err != nil {
+		t.Fatalf("create test data failed: %v", err)
+	}
+	// delete a row
+	err = db.Where("hostname = ?", "soft2.local").Delete(&Database{}).Error
+	if err != nil {
+		t.Fatalf("soft delete failed: %v", err)
+	}
+	// select non-deleted rows
+	var rows []Database
+	err = db.Find(&rows).Error
+	if err != nil {
+		t.Fatalf("find non-deleted failed: %v", err)
+	}
+	if len(rows) != 2 {
+		t.Fatalf("expected 2 non-deleted rows, got %d", len(rows))
+	}
+	for _, r := range rows {
+		if r.Hostname == "soft2.local" {
+			t.Fatalf("soft-deleted record still found in normal query")
+		}
+	}
+}
+
 func TestGormComplexQueries(t *testing.T) {
 	db := openGormDB(t)
 	err := db.AutoMigrate(&Database{})
@@ -355,4 +391,95 @@ func TestGormLastMethod(t *testing.T) {
 	if database.Hostname != "last-test-2.local" {
 		t.Fatalf("unexpected hostname from Last(): %s", database.Hostname)
 	}
+}
+
+func TestGormPartialIndexes(t *testing.T) {
+	db := openGormDB(t)
+
+	sqlDB, _ := db.DB()
+
+	_, err := sqlDB.Exec(`
+		CREATE TABLE partial_index_test (
+			id INTEGER PRIMARY KEY,
+			status TEXT,
+			priority INTEGER,
+			deleted_at TIMESTAMP,
+			email TEXT,
+			active BOOLEAN
+		)
+	`)
+	if err != nil {
+		t.Fatalf("failed to create table: %v", err)
+	}
+
+	// Test creating partial indexes
+	t.Run("CreatePartialIndexes", func(t *testing.T) {
+		_, err := sqlDB.Exec(`
+			CREATE UNIQUE INDEX idx_active_email 
+			ON partial_index_test(email) 
+			WHERE active = 1
+		`)
+		if err != nil {
+			t.Fatalf("failed to create partial unique index: %v", err)
+		}
+
+		_, err = sqlDB.Exec(`
+			CREATE INDEX idx_high_priority 
+			ON partial_index_test(priority, status) 
+			WHERE priority > 5 AND status != 'archived'
+		`)
+		if err != nil {
+			t.Fatalf("failed to create partial index with complex WHERE: %v", err)
+		}
+
+		_, err = sqlDB.Exec(`
+			CREATE INDEX idx_not_deleted 
+			ON partial_index_test(status) 
+			WHERE deleted_at IS NULL
+		`)
+		if err != nil {
+			t.Fatalf("failed to create partial index with IS NULL: %v", err)
+		}
+	})
+
+	t.Run("PartialUniqueConstraint", func(t *testing.T) {
+		// Insert active user with email
+		_, err := sqlDB.Exec(`
+			INSERT INTO partial_index_test (email, active, status, priority) 
+			VALUES (?, ?, ?, ?)
+		`, "user@example.com", true, "active", 3)
+		if err != nil {
+			t.Fatalf("failed to insert first active user: %v", err)
+		}
+
+		// Should succeed - same email but inactive user
+		_, err = sqlDB.Exec(`
+			INSERT INTO partial_index_test (email, active, status, priority) 
+			VALUES (?, ?, ?, ?)
+		`, "user@example.com", false, "inactive", 2)
+		if err != nil {
+			t.Fatalf("failed to insert inactive user with same email: %v", err)
+		}
+
+		// Should fail - duplicate email for active user
+		_, err = sqlDB.Exec(`
+			INSERT INTO partial_index_test (email, active, status, priority) 
+			VALUES (?, ?, ?, ?)
+		`, "user@example.com", true, "active", 5)
+		if err == nil {
+			t.Fatal("expected unique constraint violation for duplicate active email")
+		}
+	})
+
+	t.Run("ComplexWhereClause", func(t *testing.T) {
+		_, err := sqlDB.Exec(`
+			CREATE INDEX IF NOT EXISTS idx_complex 
+			ON partial_index_test(id) 
+			WHERE (priority BETWEEN 3 AND 7) AND status IN ('active', 'pending')
+		`)
+		if err != nil {
+			t.Fatalf("failed to create complex partial index: %v", err)
+		}
+
+	})
 }

--- a/gorm_test.go
+++ b/gorm_test.go
@@ -1,0 +1,358 @@
+package turso_go_test
+
+import (
+	"database/sql"
+	"fmt"
+	"testing"
+	"time"
+
+	_ "github.com/tursodatabase/turso-go"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+type Database struct {
+	ID               uint `gorm:"primaryKey"`
+	CreatedAt        time.Time
+	UpdatedAt        time.Time
+	DeletedAt        gorm.DeletedAt `gorm:"index"`
+	Hostname         string         `gorm:"unique;not null"`
+	Namespace        string
+	Address          string
+	PrimaryAddress   string
+	CloudClusterName string
+	Local            bool
+	AllowedIPs       string
+}
+
+func openGormDB(t *testing.T) *gorm.DB {
+	t.Helper()
+
+	sqlDB, err := sql.Open("turso", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open sql connection: %v", err)
+	}
+	db, err := gorm.Open(sqlite.Dialector{
+		Conn: sqlDB,
+	}, &gorm.Config{})
+	if err != nil {
+		t.Fatalf("failed to open gorm connection: %v", err)
+	}
+
+	t.Cleanup(func() {
+		sqlDB, _ := db.DB()
+		sqlDB.Close()
+	})
+
+	return db
+}
+
+func TestGormBasicOperations(t *testing.T) {
+	db := openGormDB(t)
+
+	// Create table
+	err := db.Debug().AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	record := Database{
+		Hostname:         "test.local",
+		Namespace:        "ns-test",
+		Address:          "http://test:8080",
+		CloudClusterName: "cluster-1",
+	}
+
+	result := db.Debug().Create(&record)
+	if result.Error != nil {
+		t.Fatalf("create failed: %v", result.Error)
+	}
+	if result.RowsAffected != 1 {
+		t.Fatalf("expected 1 row affected, got %d", result.RowsAffected)
+	}
+	if record.ID == 0 {
+		t.Fatal("expected ID to be set after create")
+	}
+
+	var found Database
+	err = db.Debug().Where("hostname = ?", "test.local").First(&found).Error
+	if err != nil {
+		t.Fatalf("find failed: %v", err)
+	}
+	if found.Hostname != "test.local" {
+		t.Fatalf("unexpected hostname: %s", found.Hostname)
+	}
+}
+
+func TestGormUpsert(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	record := Database{
+		Hostname:  "upsert-test.local",
+		Namespace: "ns-1",
+		Address:   "http://addr1",
+	}
+
+	err = db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "hostname"}},
+		UpdateAll: true,
+	}).Create(&record).Error
+
+	if err != nil {
+		t.Fatalf("first upsert failed: %v", err)
+	}
+
+	originalID := record.ID
+	if originalID == 0 {
+		t.Fatal("expected ID to be set after upsert")
+	}
+
+	record2 := Database{
+		Hostname:  "upsert-test.local",
+		Namespace: "ns-2",
+		Address:   "http://addr2",
+	}
+
+	err = db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "hostname"}},
+		UpdateAll: true,
+	}).Create(&record2).Error
+
+	if err != nil {
+		t.Fatalf("second upsert (update) failed: %v", err)
+	}
+
+	var count int64
+	db.Model(&Database{}).Where("hostname = ?", "upsert-test.local").Count(&count)
+	if count != 1 {
+		t.Fatalf("expected 1 record, got %d", count)
+	}
+
+	var updated Database
+	db.Where("hostname = ?", "upsert-test.local").First(&updated)
+	if updated.Namespace != "ns-2" {
+		t.Fatalf("expected namespace to be updated to ns-2, got %s", updated.Namespace)
+	}
+}
+
+func TestGormUpsertWithReturning(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	t.Run("RawSQLReturning", func(t *testing.T) {
+		sqlDB, _ := db.DB()
+
+		const query = `
+		INSERT INTO databases (created_at, updated_at, hostname, namespace, address, primary_address, cloud_cluster_name, local, allowed_ips)
+		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
+		ON CONFLICT (hostname) DO UPDATE SET
+			updated_at = excluded.updated_at,
+			namespace = excluded.namespace
+		RETURNING id`
+
+		stmt, err := sqlDB.Prepare(query)
+		if err != nil {
+			t.Fatalf("prepare failed: %v", err)
+		}
+		defer stmt.Close()
+
+		now := time.Now()
+		var returnedID int64
+		err = stmt.QueryRow(
+			now, now, "raw-test.local", "ns-raw",
+			"http://raw", "", "cluster-raw", false, "",
+		).Scan(&returnedID)
+
+		if err != nil {
+			t.Fatalf("raw upsert with returning failed: %v", err)
+		}
+		if returnedID == 0 {
+			t.Fatal("expected non-zero ID from RETURNING")
+		}
+		t.Logf("Raw SQL RETURNING worked, ID: %d", returnedID)
+	})
+
+	t.Run("BatchUpsert", func(t *testing.T) {
+		records := []Database{
+			{Hostname: "batch1.local", Namespace: "ns1"},
+			{Hostname: "batch2.local", Namespace: "ns2"},
+			{Hostname: "batch3.local", Namespace: "ns3"},
+		}
+
+		err := db.Debug().Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "hostname"}},
+			UpdateAll: true,
+		}).Create(&records).Error
+		if err != nil {
+			t.Fatalf("batch upsert failed: %v", err)
+		}
+
+		for i, r := range records {
+			if r.ID == 0 {
+				t.Errorf("record %d has zero ID after batch upsert", i)
+			}
+		}
+	})
+}
+
+func TestGormComplexQueries(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	testData := []Database{
+		{Hostname: "host1.local", Namespace: "ns1", UpdatedAt: time.Now().Add(-2 * time.Hour)},
+		{Hostname: "host2.local", Namespace: "ns2", UpdatedAt: time.Now().Add(-1 * time.Hour)},
+		{Hostname: "host3.local", Namespace: "ns3", UpdatedAt: time.Now()},
+	}
+
+	for _, d := range testData {
+		db.Create(&d)
+	}
+	db.Where("hostname = ?", "host2.local").Delete(&Database{})
+
+	t.Run("GetNonDeleted", func(t *testing.T) {
+		rows, err := db.Debug().Model(&Database{}).Where("deleted_at IS NULL").Rows()
+		if err != nil {
+			t.Fatalf("get non-deleted failed: %v", err)
+		}
+		defer rows.Close()
+
+		count := 0
+		for rows.Next() {
+			var database Database
+			err := db.ScanRows(rows, &database)
+			if err != nil {
+				t.Fatalf("scan rows failed: %v", err)
+			}
+			count++
+		}
+		// Should find host1 and host3
+		if count != 2 {
+			t.Fatalf("expected 2 non-deleted records, got %d", count)
+		}
+	})
+
+	t.Run("UnscopedQuery", func(t *testing.T) {
+		var count int64
+		db.Unscoped().Model(&Database{}).Count(&count)
+		if count != 3 {
+			t.Fatalf("expected 3 total records (including soft-deleted), got %d", count)
+		}
+	})
+}
+
+func TestGormColumnCountIssue(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	t.Run("UpsertReturningColumnCount", func(t *testing.T) {
+		record := Database{
+			Hostname: "column-test.local",
+		}
+
+		err := db.Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "hostname"}},
+			UpdateAll: true,
+		}).Create(&record).Error
+
+		if err != nil {
+			t.Fatalf("first upsert failed: %v", err)
+		}
+
+		record2 := Database{
+			Hostname:         "column-test.local",
+			Namespace:        "ns-test",
+			Address:          "http://test",
+			PrimaryAddress:   "http://primary",
+			CloudClusterName: "cluster",
+			Local:            true,
+			AllowedIPs:       "192.168.1.0/24",
+		}
+
+		err = db.Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "hostname"}},
+			UpdateAll: true,
+		}).Create(&record2).Error
+
+		if err != nil {
+			// This is where we expect to catch the column mismatch error
+			t.Logf("Caught potential column mismatch error: %v", err)
+			t.Fatalf("second upsert failed: %v", err)
+		}
+	})
+
+	t.Run("ExplicitReturningClause", func(t *testing.T) {
+		sqlDB, _ := db.DB()
+
+		testCases := []struct {
+			name      string
+			returning string
+		}{
+			{"returning_id", "RETURNING id"},
+			{"returning_star", "RETURNING *"},
+			{"returning_multiple", "RETURNING id, updated_at, hostname"},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				query := `
+				INSERT INTO databases (hostname, namespace)
+				VALUES (?, ?)
+				ON CONFLICT (hostname) DO UPDATE SET
+					namespace = excluded.namespace
+				` + tc.returning
+
+				rows, err := sqlDB.Query(query, tc.name+".local", "ns-test")
+				if err != nil {
+					t.Fatalf("%s query failed: %v", tc.name, err)
+				}
+				defer rows.Close()
+
+				cols, err := rows.Columns()
+				if err != nil {
+					t.Fatalf("get columns failed: %v", err)
+				}
+				t.Logf("%s returned %d columns: %v", tc.name, len(cols), cols)
+			})
+		}
+	})
+}
+
+func TestGormLastMethod(t *testing.T) {
+	db := openGormDB(t)
+	err := db.AutoMigrate(&Database{})
+	if err != nil {
+		t.Fatalf("automigrate failed: %v", err)
+	}
+
+	for i := 1; i <= 3; i++ {
+		db.Create(&Database{
+			Hostname:  fmt.Sprintf("last-test-%d.local", i),
+			Namespace: fmt.Sprintf("ns-%d", i),
+		})
+		time.Sleep(10 * time.Millisecond)
+	}
+	var database Database
+	err = db.Where(Database{Hostname: "last-test-2.local"}).Last(&database).Error
+	if err != nil {
+		t.Fatalf("Last() failed: %v", err)
+	}
+
+	if database.Hostname != "last-test-2.local" {
+		t.Fatalf("unexpected hostname from Last(): %s", database.Hostname)
+	}
+}

--- a/rows.go
+++ b/rows.go
@@ -28,10 +28,7 @@ func newRows(ctx uintptr) *tursoRows {
 }
 
 func (r *tursoRows) isClosed() bool {
-	if r.ctx == 0 || r.closed {
-		return true
-	}
-	return false
+	return r.ctx == 0 || r.closed
 }
 
 func dequoteIdent(s string) string {
@@ -66,31 +63,29 @@ func (r *tursoRows) Columns() []string {
 	if r.isClosed() {
 		return nil
 	}
-	if r.columns == nil {
-		r.mu.Lock()
-		count := rowsGetColumns(r.ctx)
-		if count > 0 {
-			cols := make([]string, 0, count)
-			for i := 0; i < int(count); i++ {
-				cstr := rowsGetColumnName(r.ctx, int32(i))
-				raw := fmt.Sprintf("%s", GoString(cstr))
-				freeCString(cstr)
-				name := dequoteIdent(baseName(raw))
-				cols = append(cols, name)
-			}
-			r.mu.Unlock()
-			r.columns = cols
-		} else {
-			r.mu.Unlock()
-		}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.columns != nil {
+		return r.columns
 	}
+	count := rowsGetColumns(r.ctx)
+	if count <= 0 {
+		return nil
+	}
+	cols := make([]string, 0, count)
+	for i := 0; i < int(count); i++ {
+		cstr := rowsGetColumnName(r.ctx, int32(i))
+		raw := GoString(cstr)
+		freeCString(cstr)
+		cols = append(cols, dequoteIdent(baseName(raw)))
+	}
+	r.columns = cols
 	return r.columns
 }
 
 func (r *tursoRows) Close() error {
-	r.err = errors.New(RowsClosedErr)
 	if r.isClosed() {
-		return r.err
+		return nil
 	}
 	r.mu.Lock()
 	r.closed = true
@@ -113,27 +108,54 @@ func (r *tursoRows) Next(dest []driver.Value) error {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	if r.isClosed() {
-		return r.err
+		if r.err != nil {
+			return r.err
+		}
+		return io.EOF
 	}
 	for {
-		status := rowsNext(r.ctx)
-		switch ResultCode(status) {
+		rc := ResultCode(rowsNext(r.ctx))
+		switch rc {
 		case Row:
-			for i := range dest {
-				valPtr := rowsGetValue(r.ctx, int32(i))
-				val := toGoValue(valPtr)
-				if val == nil {
-					r.getError()
+			ncol := int(rowsGetColumns(r.ctx))
+			if ncol < 0 {
+				if e := r.getError(); e != nil {
+					r.err = e
+				} else {
+					r.err = errors.New("rows: negative column count")
 				}
-				dest[i] = val
+				return r.err
+			}
+			if len(dest) > ncol {
+				dest = dest[:ncol]
+			}
+			for i := 0; i < len(dest); i++ {
+				vp := rowsGetValue(r.ctx, int32(i))
+				if vp == 0 {
+					if e := r.getError(); e != nil {
+						r.err = e
+					} else {
+						r.err = fmt.Errorf("rows: missing value at column %d", i)
+					}
+					return r.err
+				}
+				dest[i] = toGoValue(vp)
 			}
 			return nil
+		case ConstraintViolation:
+			r.err = errors.New("constraint violation")
+			return r.err
 		case Io:
 			continue
 		case Done:
 			return io.EOF
 		default:
-			return r.getError()
+			if e := r.getError(); e != nil {
+				r.err = e
+			} else {
+				r.err = fmt.Errorf("query failed: %s", rc.String())
+			}
+			return r.err
 		}
 	}
 }

--- a/rs_src/rows.rs
+++ b/rs_src/rows.rs
@@ -6,18 +6,29 @@ use std::{
     ffi::{c_char, c_void},
     sync::{Arc, Mutex},
 };
-use turso_core::{LimboError, Statement, StepResult, Value};
+use turso_core::{LimboError, Statement, StepResult};
 
 pub struct TursoRows<'conn> {
     stmt: Arc<Mutex<Statement>>,
+    state: RowsState,
     _conn: &'conn mut TursoConn,
     err: Option<LimboError>,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[repr(C)]
+enum RowsState {
+    Init,
+    AtRow,
+    Done,
+    Error,
 }
 
 impl<'conn> TursoRows<'conn> {
     pub fn new(stmt: Arc<Mutex<Statement>>, conn: &'conn mut TursoConn) -> Self {
         TursoRows {
             stmt,
+            state: RowsState::Init,
             _conn: conn,
             err: None,
         }
@@ -37,7 +48,7 @@ impl<'conn> TursoRows<'conn> {
 
     fn get_error(&mut self) -> *const c_char {
         if let Some(err) = &self.err {
-            let err = format!("{err}");
+            let err = err.to_string();
             let c_str = std::ffi::CString::new(err).unwrap();
             self.err = None;
             c_str.into_raw() as *const c_char
@@ -50,27 +61,30 @@ impl<'conn> TursoRows<'conn> {
 #[unsafe(no_mangle)]
 pub extern "C" fn rows_next(ctx: *mut c_void) -> ResultCode {
     if ctx.is_null() {
-        tracing::error!("rows_next: context is null");
         return ResultCode::Error;
     }
     let ctx = TursoRows::from_ptr(ctx);
     if let Ok(mut stmt) = ctx.stmt.lock() {
         match stmt.step() {
-            Ok(StepResult::Row) => ResultCode::Row,
-            Ok(StepResult::Done) => ResultCode::Done,
-            Ok(StepResult::IO) => {
-                let res = stmt.run_once();
-                if res.is_err() {
-                    ResultCode::Error
-                } else {
-                    ResultCode::Io
-                }
+            Ok(StepResult::Row) => {
+                ctx.state = RowsState::AtRow;
+                ResultCode::Row
             }
+            Ok(StepResult::Done) => {
+                ctx.state = RowsState::Done;
+                ResultCode::Done
+            }
+            Ok(StepResult::IO) => ResultCode::Io,
             Ok(StepResult::Busy) => ResultCode::Busy,
             Ok(StepResult::Interrupt) => ResultCode::Interrupt,
             Err(err) => {
+                let code = match err {
+                    LimboError::Constraint(..) => ResultCode::ConstraintViolation,
+                    _ => ResultCode::Error,
+                };
                 ctx.err = Some(err);
-                ResultCode::Error
+                ctx.state = RowsState::Error;
+                code
             }
         }
     } else {
@@ -84,13 +98,15 @@ pub extern "C" fn rows_get_value(ctx: *mut c_void, col_idx: usize) -> *const c_v
         return std::ptr::null();
     }
     let ctx = TursoRows::from_ptr(ctx);
+    if ctx.state != RowsState::AtRow {
+        return std::ptr::null();
+    }
 
     #[allow(clippy::collapsible_if)]
     if let Ok(stmt) = ctx.stmt.lock() {
         if let Some(row) = stmt.row() {
-            if let Ok(value) = row.get::<&Value>(col_idx) {
-                return TursoValue::from_db_value(value).to_ptr();
-            }
+            let res = row.get_value(col_idx);
+            return TursoValue::from_db_value(res).to_ptr();
         }
     }
     std::ptr::null()
@@ -160,6 +176,7 @@ pub extern "C" fn rows_close(ctx: *mut c_void) {
         if let Ok(mut stmt) = rows.stmt.lock() {
             stmt.reset();
             rows.err = None;
+            rows.state = RowsState::Init;
         }
     }
     unsafe {

--- a/rs_src/statement.rs
+++ b/rs_src/statement.rs
@@ -106,6 +106,7 @@ pub extern "C" fn stmt_execute(
 #[unsafe(no_mangle)]
 pub extern "C" fn stmt_parameter_count(ctx: *mut c_void) -> i32 {
     if ctx.is_null() {
+        tracing::error!("stmt_parameter_count: context is null");
         return -1;
     }
     let stmt = TursoStatement::from_ptr(ctx);
@@ -155,8 +156,7 @@ pub struct TursoStatement<'conn> {
 #[unsafe(no_mangle)]
 pub extern "C" fn stmt_close(ctx: *mut c_void) -> ResultCode {
     if !ctx.is_null() {
-        let stmt = unsafe { Box::from_raw(ctx as *mut TursoStatement) };
-        drop(stmt);
+        let _ = unsafe { Box::from_raw(ctx as *mut TursoStatement) };
         return ResultCode::Ok;
     }
     ResultCode::Invalid

--- a/stmt.go
+++ b/stmt.go
@@ -81,7 +81,10 @@ func (ls *tursoStmt) Exec(args []driver.Value) (driver.Result, error) {
 	case Invalid:
 		return nil, errors.New("invalid statement")
 	default:
-		return nil, ls.getError()
+		if e := ls.getError(); e != nil {
+			return nil, e
+		}
+		return nil, fmt.Errorf("execute failed: %s", ResultCode(res).String())
 	}
 }
 
@@ -99,7 +102,10 @@ func (ls *tursoStmt) Query(args []driver.Value) (driver.Rows, error) {
 	defer ls.mu.Unlock()
 	rowsPtr := stmtQuery(ls.ctx, argPtr, uint64(len(queryArgs)))
 	if rowsPtr == 0 {
-		return nil, ls.getError()
+		if e := ls.getError(); e != nil {
+			return nil, e
+		}
+		return nil, fmt.Errorf("query failed: no row handle")
 	}
 	return newRows(rowsPtr), nil
 }

--- a/types.go
+++ b/types.go
@@ -4,6 +4,7 @@ import (
 	"database/sql/driver"
 	"fmt"
 	"runtime"
+	"strings"
 	"time"
 	"unsafe"
 )
@@ -90,6 +91,31 @@ const (
 	FfiFreeBlob           = "free_blob"
 )
 
+var timeLayouts = []string{
+	time.RFC3339Nano,
+	time.RFC3339,
+	"2006-01-02",
+	"2006-01-02 15:04:05",
+	"2006-01-02 15:04:05Z07:00",
+	"2006-01-02 15:04:05.999999999",
+	"2006-01-02 15:04:05.999999999Z07:00",
+}
+
+func parseTimeString(str string) (time.Time, bool) {
+	if t, err := time.Parse(time.RFC3339Nano, str); err == nil {
+		return t, true
+	}
+	for _, layout := range timeLayouts[1:] {
+		if t, err := time.Parse(layout, str); err == nil {
+			if !strings.Contains(layout, "Z07:00") && !strings.Contains(layout, "T") {
+				return t.In(time.UTC), true
+			}
+			return t, true
+		}
+	}
+	return time.Time{}, false
+}
+
 // convert a namedValue slice into normal values until named parameters are supported
 func namedValueToValue(named []driver.NamedValue) []driver.Value {
 	out := make([]driver.Value, len(named))
@@ -160,9 +186,10 @@ func toGoValue(valPtr uintptr) any {
 		defer freeCString(textPtr)
 		str := GoString(textPtr)
 
-		// Try to parse as RFC3339 time format
-		if t, err := time.Parse(time.RFC3339, str); err == nil {
-			return t
+		if strings.Contains(str, "T") || (strings.Contains(str, " ") && strings.Contains(str, ":")) {
+			if t, ok := parseTimeString(str); ok {
+				return t
+			}
 		}
 
 		// If it doesn't parse as time, return as string

